cached_property.py

```diff
@@ -14,54 +15,104 @@
     asyncio = None
 
 
+def _is_mangled(name):
+    """Check whether a name will be mangled when accessed as an attribute."""
+    return name.startswith("__") and not name.endswith("__")
+
+
+def _resolve_name(klass, default=None):
+    """Get the name of a function used to access it from an object."""
+    name = default
+
+    if _is_mangled(name) and klass:
+        for base_cls in inspect.getmro(klass):
+            resolved_name = "_{}{}".format(base_cls.__name__, name)
+            if resolved_name in base_cls.__dict__:
+                name = resolved_name
+                break
+
+    return name
+
+
 class cached_property(object):
     """
     A property that is only computed once per instance and then replaces itself
     with an ordinary attribute. Deleting the attribute resets the property.
     Source: https://github.com/bottlepy/bottle/commit/fa7733e075da0d790d809aa3d2f53071897e6f76
     """  # noqa
 
-    def __init__(self, func):
-        self.__doc__ = getattr(func, "__doc__")
-        self.func = func
+    def __init__(self, func, name=None):
+        if not callable(func):
```

Comment:
I wasn't sure if this was the right thing to do here. I did consider making `func` optional so that users have to specify arguments with keywords. Ultimately I decided to go with the pattern that already existed in the other classes.